OrbitGl/App.cpp

```diff
@@ -839,32 +832,71 @@ void OrbitApp::SendErrorToUi(const std::string& title,
-  GetSymbolsClient()->LoadSymbolsFromModules(Capture::GTargetProcess.get(),
-                                             m_ModulesToLoad, nullptr);
-  m_ModulesToLoad.clear();
-  // This is a bit counterintuitive. LoadSymbols generates request to
-  // the service if symbols cannot be loaded locally. In which case
-  // the UI update happens in OnRemoteModuleDebugInfo. But if all symbols
-  // are loaded locally it will not call remote and OnRemoteModuleDebugInfo
-  // are never called. For this case we need to make sure sampling report is
-  // updated here as well.
-  // TODO: LoadSymbolsFromModules should always call a callback even when
-  // symbols are loaded locally. Remove this call after this is done.
-  UpdateSamplingReport();
-  GOrbitApp->FireRefreshCallbacks();
+void OrbitApp::LoadRemoteModules(uint32_t process_id,
+                                 const std::shared_ptr<Session>& session) {
+  if (modules_to_load_.empty()) return;
+
+  std::shared_ptr<Module> module = *modules_to_load_.begin();
+  modules_to_load_.erase(module);
+  modules_currently_loading_.insert(module);
+
+  thread_pool_->Schedule([this, module, process_id, session]() {
+    std::string error_str;
+    if (symbol_helper_.LoadSymbolsUsingSymbolsFile(module)) {
+      LOG("Loaded %lu function symbols locally for modules %s",
+          module->m_Pdb->GetFunctions().size(), module->m_Name);
+    } else {
+      LOG("Did not find local symbols for module: %s", module->m_Name);
+      const auto remote_symbols_result =
+          process_manager_->GetSymbols(process_id, module->m_Name);
+      if (!remote_symbols_result) {
+        error_str = remote_symbols_result.error();
+      } else {
+        symbol_helper_.LoadSymbolsIntoModule(module,
+                                             remote_symbols_result.value());
+        LOG("Received and loaded %lu function symbols from remote service "
+            "for module %s",
+            module->m_Pdb->GetFunctions().size(), module->m_Name.c_str());
+      }
+    }
+
+    if (error_str.empty() && session != nullptr &&
+        session->m_Modules.find(module->m_Name) != session->m_Modules.end()) {
+      module->m_Pdb->ApplyPresets(*session);
+    }
+
+    main_thread_executor_->Schedule([this, module,
+                                     error_str = std::move(error_str),
+                                     process_id, session]() {
+      modules_currently_loading_.erase(module);
+      if (!error_str.empty()) {
+        SendErrorToUi(
+            "Error loading symbols",
+            absl::StrFormat(
+                "Did not find symbols on local machine for module %s. "
+                "Trying to load symbols from remote resulted in error "
+                "message: %s",
+                module->m_Name, error_str));
+      } else {
+        UpdateSamplingReport();
+        GOrbitApp->FireRefreshCallbacks();
+      }
+      // Load next module
+      LoadRemoteModules(process_id, session);
```

Comment:
I'm a bit confused by the recursive call instead of a loop. If there's a specific reason please add a comment.

Reply:
I had a reason for that, but I rewrote it to use a loop as before